### DISCUSSION TITLE
How to forward defer's setCookie to the client response header in Remix

### DISCUSSION THREAD

**Discussion by 19Qingfeng** [remix-run/remix#10218] [Q&A]
I need to forward all response headers to the client for effectiveness after the request in defer is returned, like this: 
```ts
import {
  defer,
  Links,
  Meta,
  Outlet,
  Scripts,
  ScrollRestoration
} from '@remix-run/react';
import type { LinksFunction, LoaderFunction } from '@remix-run/node';

export const loader: LoaderFunction = () => {
  async function messagePromise() {
    // How to use https://api.example.com/messages Forward the returned responsiveHeader to the client
    return fetch('https://api.example.com/messages');
  }

  return defer({
    messagePromise
  });
};

export function Layout({ children }: { children: React.ReactNode }) {
  const { messagePromise } = useLoaderData();

  return (
    <html lang="en">
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <Meta />
        <Links />
      </head>
      <body>
        {children}
        <ScrollRestoration />
        <Scripts />
      </body>
    </html>
  );
}

export default function App() {
  return <Outlet />;
}
```

How to use https://api.example.com/messages Forward the returned responsiveHeader to the client?

Is there any mechanism or API in Remix that can be implemented

**Comment by akomm** [comment]
Doesn't seem to make any sense to me, but:

You do know, that `fetch` returns a `Promise<Response>` and you can directly return it from the loader? No defer needed. It will also properly stream / pass-through the data as a stream without blocking. And obviously contain all the headers from the fetch Response. 


Fetch returns a Response promise:
![grafik](https://github.com/user-attachments/assets/f685c6ac-2ce6-4d94-a106-02a3f6540625)

Which provides you the headers...

There is potentially also a way to do it with defer, if you need to make multiple fetch and return multiple results. But at this point what do you do with the multiple headers?

it seems weird to me on many levels.

**Reply by akomm** [reply]
If you need defer and headers, you need to do something like that:

![grafik](https://github.com/user-attachments/assets/dddaa2f0-7897-47e1-a6d6-77f08f7f5e55)

But it might be not very efficient.

**Reply by akomm** [reply]
when fetch resolves only when body is fully streamed, you won't gain anything. If fetch resolves when headers are received and body starts stream, you save the time waiting for the body to arrive fully.

**Reply by 19Qingfeng** [reply]
> when fetch resolves only when body is fully streamed, you won't gain anything. If fetch resolves when headers are received and body starts stream, you save the time waiting for the body to arrive fully.

In fact, we will filter and synchronize specific cookies on the server according to the order of each request. Simply put, for example, we will filter the same key in chronological order

**Reply by 19Qingfeng** [reply]
> If you need defer and headers, you need to do something like that:
> 
> ![grafik](https://private-user-images.githubusercontent.com/2409779/387287179-dddaa2f0-7897-47e1-a6d6-77f08f7f5e55.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM1NzMxMDAsIm5iZiI6MTczMzU3MjgwMCwicGF0aCI6Ii8yNDA5Nzc5LzM4NzI4NzE3OS1kZGRhYTJmMC03ODk3LTQ3ZTEtYTZkNi03N2YwOGY3ZjVlNTUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMDdUMTIwMDAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGFlZTUzYmQ1MzRjYmVjYWQwOWZkNmNjNTRlNjRhMjZiYmFmYmUwOWMzODVjZTU3NDQ3Zjg2NmQ2YzkwNmRmYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.hCZ5ahl46Twxi9QcZhPUA5wzetjNRRuE5kRCkB24YkM)
> 
> But it might be not very efficient.

It seems that this is currently the way for Remix to use both defer and synchronize server API response headers to the client without any modifications.

I will try to see if this behavior can meet our demands. Thank you very much for your answer.

**Comment by sergiodxa** [comment]
The headers are sent on the response before the body

Defer allows you to send a response and stream the body later 

If you need some data to set the headers then you need the data before you send the response

This means you need to await that data and can't defer it anymore

If the data is not required to set a header or status code you can defer it

**Reply by akomm** [reply]
From post it appears they need the headers from fetch response to pass down the loader. So if fetch resolves response before body is streamed, the headers are available directly and can be forwarded like in my example to defer. Response.json() will collect the body stream data and when "end" parse it as json. So returning the  promise to defer would not block.

**Reply by 19Qingfeng** [reply]
> The headers are sent on the response before the body
> 
> Defer allows you to send a response and stream the body later
> 
> If you need some data to set the headers then you need the data before you send the response
> 
> This means you need to await that data and can't defer it anymore
> 
> If the data is not required to set a header or status code you can defer it


I understand what you mean, or rather we hope that when using defer, we can also pass the server-side header to the client through defer.

Because we do need this scenario, a large number of defers will bring significant performance improvements to our pages. But at this moment, the API we call on the server cannot synchronize cookies to the client.

At present, our solution can be roughly divided into two ideas: 

1. Obtaining the cookie for the server-side API in the responder of defer, and then re calling the setCookie API on the client side after obtaining it, is unreliable both in terms of transmission and httponly security. It also has many limitations, such as the loss of the httponly property.

2. We would like to attempt to initiate a retry request on the client side after all defer requests have been completed, that is, after all client promises have been fulfilled. The server will issue a response header for all AKS ookies in this defer request based on the current refetch request.

It seems that the second method is indeed feasible, and I think it is a more universal approach.

We have tried to add him to Remix, but I don't know how the Remix team views such behavior. Perhaps it can be integrated into the default framework as a configuration item, providing users with the freedom to choose whether to enable it or not.

**Reply by akomm** [reply]
It sounds very weird and more like some root problem that should be found and addressed rather than implementing awkward solutions.

It should not make a difference whether you navigate and interact client side or full reload the page, but with this method you will have race conditions, stale UI and hard to debug issues.

If you fetch data, you shouldn't set cookies.
